### ISSUE TITLE
Self Signed Certificate on On-Prem Gitlab generates error about common name with http backend

### ISSUE THREAD

**Issue by MichaelSweikata** [hashicorp/terraform#28371]
### Terraform Version

```
2021/04/14 21:07:17 [INFO] Terraform version: 0.14.10  
2021/04/14 21:07:17 [INFO] Go runtime version: go1.15.6
```

And with the release of 0.15 today, I've even tried that:
```
bash-5.1# terraform -version
Terraform v0.15.0
on linux_amd64
```

### Terraform Configuration Files
_Backend file:_
```terraform
terraform {
    backend "http" { }
}
```

_backend tfvars file:_
```terraform
lock_method = "POST" 
unlock_method = "DELETE" 
retry_max = "3" 
skip_cert_verification = true
```

### Debug Output
https://gist.github.com/MichaelSweikata/eebfe527831e5811490eeae24e2c2451

### Expected Behavior
The certificate verification should have been skipped.

### Actual Behavior
It generated an error and stopped the authentication process

### Steps to Reproduce
Running the init from the gitlab CI container

```bash
terraform -chdir="./app_teams/${team_folder}" init -reconfigure -backend-config="address=https://gitlab.buildone.co/api/v4/projects/${CI_PROJECT_ID}/terraform/state/${team_folder}" -backend-config="lock_address=https://gitlab.buildone.co/api/v4/projects/${CI_PROJECT_ID}/terraform/state/${team_folder}/lock" -backend-config="unlock_address=https://gitlab.buildone.co/api/v4/projects/${CI_PROJECT_ID}/terraform/state/${team_folder}/lock" -backend-config="username=project_${CI_PROJECT_ID}_bot" -backend-config="password=${bot_2_token}"
```

### Additional Context
I'm running this in a gitlab pipeline in a docker image from an alpine image I created and loaded the self-signed cert into, via the CLI.

I've also tried including some default values in the backend "http" {} code instead of a tfvars, and that just generated the generic 401 error:

```terraform
2021/04/14 20:48:35 [DEBUG] GET https://gitlab.buildone.co/api/v4/projects/2/terraform/state/infrastructure
Successfully configured the backend "http"! Terraform will automatically
use this backend unless the backend configuration changes.
Error refreshing state: HTTP remote state endpoint requires auth
```

Which is why in this attempt, I left the backend http empty, and just put the static values in the tfvars, and the dynamic values as backend-configs, or attempted them with the environment shell.

### References
None that I could find.

**Comment by MichaelSweikata** [comment]
I've even deployed my self signed cert to include SAN values that match the server name, repulled them into the container, and if I still do the same command sets, I now just get:

`Error refreshing state: HTTP remote state endpoint requires auth`

**Comment by apparentlymart** [comment]
HI @MichaelSweikata,

From reviewing the `http` backend code, it looks like "HTTP remote state endpoint requires auth" is the message used to report that the remote API returned the `401 Unauthorized` status code:

https://github.com/hashicorp/terraform/blob/2cd1619c40124116cc65350c2c321479ce5237b9/backend/remote-state/http/client.go#L95-L96

Since that's an error related to the HTTP _response_, I wouldn't expect it to be caused by a TLS certificate verification error; Terraform wouldn't even be able to send the HTTP _request_ if the TLS handshake were failing.

I'm not personally familiar with this GitLab state API, but my first guess here would be that the credentials you've provided here are not correct, and so the GitLab server is returning an authentication error. Perhaps you could confirm that by trying a request directly to that API from outside of Terraform, using the same credentials, to see if it returns the same error.

**Comment by MichaelSweikata** [comment]
See, that's what I thought when I looked at the repo. But when I test the token manually, I get a 200.

```
bash-5.1# curl -k -I "https://gitlab.buildone.co/api/v4/projects?eq-Xxhs-MDyCNDUDMHos"
HTTP/2 200 
server: nginx
date: Wed, 14 Apr 2021 23:59:43 GMT
content-type: application/json
vary: Accept-Encoding
cache-control: no-cache
link: <https://gitlab.buildone.co/api/v4/projects?eq-Xxhs-MDyCNDUDMHos=&membership=false&order_by=created_at&owned=false&page=1&per_page=20&repository_checksum_failed=false&simple=false&sort=desc&starred=false&statistics=false&wiki_checksum_failed=false&with_custom_attributes=false&with_issues_enabled=false&with_merge_requests_enabled=false>; rel="first", <https://gitlab.buildone.co/api/v4/projects?eq-Xxhs-MDyCNDUDMHos=&membership=false&order_by=created_at&owned=false&page=1&per_page=20&repository_checksum_failed=false&simple=false&sort=desc&starred=false&statistics=false&wiki_checksum_failed=false&with_custom_attributes=false&with_issues_enabled=false&with_merge_requests_enabled=false>; rel="last"
vary: Origin
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
x-gitlab-feature-category: projects
x-next-page:
x-page: 1
x-per-page: 20
x-prev-page:
x-request-id: 01F39D73G8ZXPJ1E1ZAK0ZS860
x-runtime: 0.184197
x-total: 1
x-total-pages: 1
strict-transport-security: max-age=31536000
referrer-policy: strict-origin-when-cross-origin

bash-5.1# 
```

So my best guess is either the method of credentialing the http request is incorrect, or there's something else under the hood I don't get. This token doesn't require a username when it accesses the API:

https://docs.gitlab.com/ee/api/README.html#personalproject-access-tokens

```bash
curl "https://gitlab.example.com/api/v4/projects?private_token=<your_access_token>"
```

Now I'm wondering if I just need to unset the TF_HTTP_USERNAME/TF_USERNAME and leave it blank?

**Comment by MichaelSweikata** [comment]
Okay, I think it's important for any viewers following or finding this in later searches.

So, there's been two problems:

I think there's an issue with Terraform and the self-signed certificate behavior (this might even be the go code underneath). Because even when I tell it to not check the cert, it still reported a very odd error that the cert was using common name, and not SAN values. But that was a symptom of another problem.

My ACTUAL problem, is that I believe the Gitlab documentation is wrong, and this wasn't an issue with Terraform once I regenerated the local cert to include the server name as a SAN value. 

First of all, @apparentlymart  , I really do appreciate your very well articulated ideas, patterns, and ideas. You've been phenomenal in helping me with this. And here's where Terraform communities can stop caring, but I'm putting it here for future browsers.

So according to the gitlab documentation here (https://docs.gitlab.com/ee/user/infrastructure/terraform_state.html#get-started-using-local-development) say, on step 4:
 
    Create a Personal Access Token with the api scope.
 
Which, instead of a personal access token which takes a license seat count, I created a Project Token: https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html
 
This token is local to the project, and has a username of

    project_$(project_id)_bot#

Which, in my case, is project_2_bot (no number since it's the first one, dumb)
 
According to docs, it says the token needs API access to hit the API. 
 
The docs are wrong. The token I created had only API access, and it kept failing. Once I gave it more access, (api, read/write to container, registry, repository), it actually went through.
